<!--
  Launches the infrastructure running on a robot (turtle) with rostful,
  as part of concert, using interactions.
 -->
<launch>
    <arg name="turtle_name"              default="turtle"/>
    <arg name="turtle_rapp_whitelist"    default="[]"/>
    <arg name="turtle_concert_whitelist" default="[]"/>
    <arg name="disable_zeroconf"         default="true"/>

    <include file="$(find rocon_app_manager)/launch/concert_client.launch">
        <arg name="robot_name" value="$(arg turtle_name)"/>
        <arg name="concert_whitelist" value="$(arg turtle_concert_whitelist)"/>
        <arg name="concert_uri"  value="http://localhost:6380"/>
        <arg name="disable_zeroconf" value="$(arg disable_zeroconf)"/>
        <arg name="robot_type" value="pc"/>
        <arg name="rapp_package_whitelist" value="$(arg turtle_rapp_whitelist)" />
        <arg name="concert_watch_period" value="1"/> <!-- flip faster, convenient -->
        <arg name="firewall" value="false"/>
        <arg name="robot_unique_name" value="false"/>  <!-- do not change this to true! the turtlesim service sets its own unique names for the robots -->
        <arg name="local_machine_only" value="true"/>
        <arg name="screen" value="true" />
        <!-- TODO : when we can pass custom rosbridge adress
        <arg name="interactions" value="true"/>
        <arg name="interactions_list" value="[rostful/rostful, rostful_examples/stroll, rostful_examples/teleop]"/>
        -->
    </include>

    <param name="name" value="Rostful Interactions"/>
    <param name="description" value="A REST interface to ROS using interactions/remocons."/>
    <param name="icon" value="rocon_icons/cybernetic_pirate.png"/>

    <arg name="rosbridge_port"           default="8880"/>
    <arg name="rosbridge_address"        default="localhost"/>

    <node pkg="rocon_interactions" type="interactions_manager.py" name="interactions">
        <rosparam param="interactions">[rostful/rostful, rostful_examples/stroll, rostful_examples/teleop]</rosparam>
        <param name="rosbridge_address" value="$(arg rosbridge_address)"/>
        <param name="rosbridge_port" value="$(arg rosbridge_port)"/>
        <param name="rapp_manager" value="true"/>
    </node>

    <include file="$(find rosbridge_server)/launch/rosbridge_websocket.launch" >
        <arg name="port" value="$(arg rosbridge_port)"/>
    </include>

    <arg name="rostful_port"             default="8080"/>

    <node pkg="rostful" name="rostful_server" type="devserver" args="-p $(arg rostful_port)">
        <param name="topics" value="['$(arg turtle_name)/rapp_list', '$(arg turtle_name)/incompatible_rapp_list', '$(arg turtle_name)/status']" type="str" />
        <param name="services" value="['interactions/get_interactions','interactions/get_roles', 'interactions/get_interaction',  'interactions/set_interaction',  'interactions/request_interaction',
                            '$(arg turtle_name)/invite', '$(arg turtle_name)/list_rapps', '$(arg turtle_name)/platform_info', '$(arg turtle_name)/start_rapp', '$(arg turtle_name)/stop_rapp',
                            'rosapi/services', 'rosapi/topics']" type="str" />
        <param name="actions" value="[]" type="str" />

        <param name="enable_rocon" value="true" type="bool" />
        <param name="rapps_namespaces" value="['$(arg turtle_name)']" type="str" />
        <param name="interactions" value="[]" type="str" />
    </node>

</launch>
